### PR TITLE
[test] nn.Pipe: add a parity test that also tests with amp

### PR DESCRIPTION
**Pull request by msbaines**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/pipe/test_parity.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    for p_pg, reg_pg in zip(popt.param_groups, ropt.param_groups):
+        for p_pg, reg_pg in zip(p_pg["params"], reg_pg["params"]):
+            assert torch.allclose(
+                p_pg.cuda(), reg_pg, atol=1e-2
+            ), f"Model parameters differ in between Pipe and Vanilla {[o_pg]} {reg_pg}"
+
+        for p_buf, reg_buf in zip(pmodel.buffers(), rmodel.buffers()):
+            assert torch.allclose(
+                p_buf.cuda(), reg_buf, atol=1e-2
+            ), "Model buffers differ in between Offload and Vanilla."
```
Comment (anj-s) [inline tests/nn/pipe/test_parity.py:45]: Update string?

File: tests/nn/pipe/test_parity.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Testing Pipe Module
```
Comment (anj-s) [inline tests/nn/pipe/test_parity.py:7]: Update string to mention partity?

File: tests/nn/pipe/test_parity.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import numpy as np
+import pytest
+import torch
+
+from fairscale.nn import Pipe
+from fairscale.utils.testing import skip_if_single_gpu
+
+
+def _get_model(num_inputs=2, num_hidden=20, num_outputs=2):
+    num_layers = torch.cuda.device_count() - 2
```
Comment (anj-s) [inline tests/nn/pipe/test_parity.py:22]: Will this always be 0? I don't remember if we run our tests on 4 GPU or only 2  on circleCI.

**Review by anj-s — APPROVED** [review]

File: tests/nn/pipe/test_parity.py (hunk 3 above)
Comment (msbaines) [reply tests/nn/pipe/test_parity.py:22]: we are using 4 GPUs

**Pull request merged**